core/src/main/scala/org/apache/spark/storage/memory/MemoryStore.scala

```diff
@@ -694,7 +694,7 @@ private[storage] class PartiallyUnrolledIterator[T](
   }
 
   override def next(): T = {
-    if (unrolled == null) {
+    if (unrolled == null || unrolled.isEmpty) {
```

Comment:
I've fixed it with the change like:

       private def readBroadcastBlock(): T = Utils.tryOrIOException {
         TorrentBroadcast.synchronized {
           setConf(SparkEnv.get.conf)
           val blockManager = SparkEnv.get.blockManager
    -      blockManager.getLocalValues(broadcastId).map(_.data.next()) match {
    -        case Some(x) =>
    -          releaseLock(broadcastId)
    -          x.asInstanceOf[T]
    +      blockManager.getLocalValues(broadcastId) match {
    +        case Some(blockResult) =>
    +          if (blockResult.data.hasNext)  {
    +            val x = blockResult.data.next().asInstanceOf[T]
    +            releaseLock(broadcastId)
    +            x
    +          } else {
    +            throw new SparkException(s"Failed to get locally stored broadcast data: $broadcastId")
    +          }